### PR TITLE
[core] support to drop partition when delete tag

### LINKED ISSUE #3041: [Feature] Support to drop partition when delete the tag
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Currently, the tag-to-partition only support create partition to hive metastore when tag is created. We should also drop the partition in metastore when delete the tag.

### Solution

Add `notifyDeletion` in `TagCallback`

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Aitozi**
### Purpose

Linked issue: close #3041

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/BatchWriteGeneratorTagOperator.java [hunk 1]
```diff
@@ -114,15 +117,15 @@ private void createTag() {
         try {
             // If the tag already exists, delete the tag
             if (tagManager.tagExists(tagName)) {
-                tagManager.deleteTag(tagName, tagDeletion, snapshotManager);
+                tagManager.deleteTag(tagName, tagDeletion, snapshotManager, callbacks);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/BatchWriteGeneratorTagOperator.java:120]: Be careful here, you need to create new callbacks, the callbacks will be closed here.

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveMetastoreClient.java [hunk 2]
```diff
@@ -91,6 +91,22 @@ public void addPartition(LinkedHashMap<String, String> partitionSpec) throws Exc
         }
     }
 
+    @Override
+    public void deletePartition(LinkedHashMap<String, String> partitionSpec) throws Exception {
+        List<String> partitionValues = new ArrayList<>(partitionSpec.values());
+        try {
+            client.getPartition(
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveMetastoreClient.java:98]: Why we need to get first?

Comment (Aitozi) [reply paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveMetastoreClient.java]: Not necessary, removed.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/BatchWriteGeneratorTagOperator.java (hunk 1 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/BatchWriteGeneratorTagOperator.java]: You are right, but I see some other same usage eg: `TagAutoCreation#callbacks`?  I think it's the same error, do i have to fix it in this PR ?

Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/BatchWriteGeneratorTagOperator.java]: Yes, you can fix them together.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**